local Functions = {}
local Vector2new = Vector2.new

Functions.lerp = function(Start, Goal, Time)
    return Start + (Goal - Start) * Time
end

local drawnew = Drawing.new
function Functions.create(type, props)
    local drawing = drawnew(type)

    for prop, value in props do
        setrenderproperty(drawing, prop, value)
    end

    return drawing
end

local EventLog = {}
local EventLogs = {}

EventLog.GlobalSize = 12
EventLog.Position = Vector2new(0, 36 * 2);
EventLog.Padding = 4

type EventLog = {
    --Priority: number;
    Duration: number;
    ClockAt: number;
    Text: string;
    Object: any?;
    Destroy: () -> ()
}

local function convertTransparency(transparency: number): number
    return math.clamp(1 - transparency, 0, 1)
end

twait = task.wait;
tspawn = task.spawn;

local White = Color3.new(1, 1, 1)
local Grey = Color3.new(.2, .2, .2)
local Black = Color3.new()
function EventLog:Log(Text, Duration)
    local Priority = #EventLogs + 1
    local Log = {}

    Log.Started = false
    Log.Priority = Priority
    Log.ClockAt = tick()
    Log.Duration = Duration
    Log.Text = Text
    Log.Object = Functions.create('Text', {
        ['Font'] = 2;
        ['Text'] = Text;
        ['Size'] = self.GlobalSize or 14;
        ['Color'] = White;
        ['Outline'] = true;
        ['OutlineColor'] = Black;
        ['Position'] = EventLog.Position + Vector2new(0, (EventLog.GlobalSize + 2) * Priority);
        ['Visible'] = true
    })

    local BoundOffset = Vector2new(5, 5)
    Log.Background = Functions.create('Square', {
        ['Size'] = Log.Object.TextBounds + BoundOffset;
        ['Color'] = Grey;
        ['Filled'] = true;
        ['ZIndex'] = -1;
        ['Visible'] = false;
    })

    Log.Background.Position = Log.Object.TextBounds + (BoundOffset / 2)

    tspawn(function()
        for time = 1, 20 do
            Log.Object.Transparency = time / 10;
            Log.Background.Transparency = (time / 10)
            twait()
        end
    end)

    function Log:Remove()
        tspawn(function()
            self.Destroyed = true
            for time = 1, 40 do
                Log.Object.Transparency = convertTransparency(time / 40)
                Log.Background.Transparency = convertTransparency(time / 40)
                twait()
            end
            EventLogs[Priority] = nil
            Log.Object:Remove()
            Log.Background:Remove()
            setmetatable(self, nil)
        end)
    end

    EventLogs[Priority] = Log
    return Log
end

local NerdSentences = { 'im such a skid ðŸ˜œ', 'math is hard ðŸ§®', 'i speak klingon ðŸ‘½', 'i wear socks with sandals ðŸ‘£', 'my pocket ðŸ˜Ž protector brings all the nerds to the yard', 'phd in quantum mechanics from hogwarts ðŸŽ“', 'im a nerd and im proud ðŸ¤“ðŸ‘Š', 'i mainframe ðŸ’» for fun', 'i dream in binary ðŸ’­', 'im fluent in c++ ðŸ’»', 'my code is poetry ðŸ“', }
local Emojis = { [1] = 'ðŸ˜Š', [2] = 'ðŸ˜¢', [3] = 'ðŸ˜‚', [4] = 'â¤ï¸', [5] = 'ðŸ‘', [6] = 'ðŸ‘Ž', [7] = 'ðŸ”¥', [8] = 'ðŸ‘', [9] = 'ðŸš€', [10] = 'ðŸ¤”', [11] = 'ðŸ˜®', [12] = 'ðŸ˜Ž', [13] = 'ðŸ˜', [14] = 'ðŸ˜¡', [15] = 'ðŸ‘»', [16] = 'ðŸ‘½', [17] = 'ðŸ¤–', [18] = 'ðŸ¦„', [19] = 'ðŸ˜º', [20] = 'ðŸ¶', [21] = 'ðŸ ', [22] = 'ðŸ•', [23] = 'ðŸŒ®', [24] = 'ðŸ°', [25] = 'â˜•', [26] = 'ðŸº', [27] = 'ðŸ·', [28] = 'ðŸ¹', [29] = 'ðŸŽµ', [30] = 'ðŸŽ¬', [31] = 'ðŸŽ®', [32] = 'âš½', [33] = 'ðŸ€', [34] = 'âš¾', [35] = 'ðŸˆ', [36] = 'ðŸŽ¾', [37] = 'â›³', [38] = 'ðŸ„', [39] = 'â›·ï¸', [40] = 'ðŸ‚', [41] = 'ðŸš´', [42] = 'ðŸŠ', [43] = 'ðŸƒ', [44] = 'ðŸ¥¾', };

local function RandomValue(Table)
    return Table[math.random(1, #Table)]
end;

game:GetService('RunService').RenderStepped:Connect(function()
    local ETC = 0
    local GlobalSize = EventLog.GlobalSize;
    local Padding = EventLog.Padding;
    local NotificationPosition = EventLog.Position;
    for Index, Log: EventLog in next, EventLogs do
        local Object = Log.Object
        local Background = Log.Background;
        if not Log.Started then
            Object.Position = NotificationPosition + Vector2new(-(Object.TextBounds.X - 2), (GlobalSize + Padding) * ETC);
            Background = Object.Position;
            Log.Started = true
        end

        ETC += 1
        Object.Position = Functions.lerp(Object.Position, NotificationPosition + Vector2new(20, (GlobalSize + Padding) * ETC), .1);
        pcall(function()
            Background.Position = Object.Position;
        end)
        if tick() - Log.ClockAt > Log.Duration then
            if not Log.Destroyed then
                Log:Remove()
            end
        end
    end
end)

return EventLog, EventLogs;
